tools/cabana/mainwin.cc

```diff
@@ -193,6 +193,27 @@ void MainWindow::createDockWindows() {
   video_dock->setFeatures(QDockWidget::DockWidgetMovable | QDockWidget::DockWidgetFloatable);
   video_dock->setWidget(video_splitter);
   addDockWidget(Qt::RightDockWidgetArea, video_dock);
+
+#ifdef WSL2
+  dock->setFeatures(QDockWidget::DockWidgetMovable);
+  video_dock->setFeatures(QDockWidget::DockWidgetMovable);
+  QObject::connect(dock, &QDockWidget::topLevelChanged, this, &MainWindow::onDockWidgetTopLevelChanged);
+  QObject::connect(video_dock, &QDockWidget::topLevelChanged, this, &MainWindow::onDockWidgetTopLevelChanged);
+#endif
+}
+
+// This is a workaround for a bug in WSL2
+void MainWindow::onDockWidgetTopLevelChanged(bool topLevel) {
+  QDockWidget *dockWidget = qobject_cast<QDockWidget *>(sender());
+  if (dockWidget) {
+    if (topLevel) {
+      // the floating button breaks mouse tracking when docked so I remove it when docked and add it back when floating
+      dockWidget->setFeatures(QDockWidget::DockWidgetMovable | QDockWidget::DockWidgetFloatable); 
+      dockWidget->setTitleBarWidget(nullptr); // this fixes the mouse tracking issue when floating
```

Comment:
1. I'm curious why setting `setTitleBarWidget(nullptr)` does the trick. the origin `titlebarWidget` is already nullptr, Calling this function just causes `QDockWidget` to reset window flags internally. am I right?
2. if it works, Can we delete other codes like set different dock features on WSL2, Just set titlebar widget to null to let qdockwidget reset window flags after toplevelchanged?

Reply:
If this issue only occurs after restarting cabana, the problem may be restoring state before showing the main window.  would a delayed call to `restoreState(settings.window_state);` fix it? : `QTimer::singleShot(0, [this](){restoreState(settings.window_state);});`

Reply:
setting setTitleBarWidget(nullptr); fixes the issue when dragging the titlebar to undock/re-dock but if I press the float button it causes the same old behaviour of not responding to events.
> This is done because pressing the floatButton will lead to the behavior of the widget not responding to any QEvents.

Reply:
the way I figured this workaround out was by looking at the qt QDockWidget and i read that there was some special behavior with the title bar. 

from https://doc.qt.io/qt-5/qdockwidget.html#QDockWidget

> If a title bar widget is set, [QDockWidget](https://doc.qt.io/qt-5/qdockwidget.html) will not use native window decorations when it is floated.